src/py2spack/package_providers.py

```diff
@@ -0,0 +1,316 @@
+
+    def get_pyproject(
+        self, name: str, version: vn.Version
+    ) -> dict[Any, Any] | PyProjectProviderQueryError:
+        """Download and extract the pyproject.toml for the specified package version."""
+        all_metadata = self._get_distribution_metadata(name)
+
+        if isinstance(all_metadata, PyProjectProviderQueryError):
+            return all_metadata
+
+        metadata = all_metadata[version]
+        # python sdist archives contain a top level directory, e.g. "black-24.4.2/"
+        directory_str = f"{name}-{version}"
+
+        # for type checker, we know these values are going to be strings
+        assert isinstance(metadata["url"], str)
+        assert isinstance(metadata["extension"], str)
+
+        return try_load_toml(metadata["url"], directory_str, metadata["extension"])
+
+    @functools.lru_cache(maxsize=1)  # noqa: B019
+    def _get_distribution_metadata(
+        self, name: str
+    ) -> dict[vn.Version, dict[str, str | dict[Any, Any]]] | PyProjectProviderQueryError:
+        """Get metadata for available distribution files from PyPI.
+
+        We cache the result in order to avoid repeatedly processing the same data
+        for all method calls to `get_pyproject` (which in turn calls this method).
+        The maxsize of 1 is due to the reasoning described in the `_get` method.
+        """
+        # NOTE: caching both the _get_distribution_metadata and the _get methods might
+        # seem redundant, idea is that the caching of _get_distribution_metadata avoids
+        # repeated processing and caching of _get makes sure the calls to get_versions
+        # and get_pyproject do not both make a call to the API.
+        # due to the cachesize of 1 for both caches, data is only stored during the con-
+        # version of the current package
+        data = self._get(name)
+        if isinstance(data, PyProjectProviderQueryError):
+            return data
+        files = data["files"]
+
+        # for now we only support tarball archives like .tar.gz
+        files_known_format = [f for f in files if _archive_format_is_known(f["filename"])]
+
+        if not len(files_known_format):
+            msg = (
+                "No files with known archive format found (note: wheel file"
+                " parsing not supported)"
+            )
+            return PyProjectProviderQueryError(msg)
+
+        # for each file, get the filename, url, version, extension, and sha256
+        # TODO @davhofer: in case of an error, skip the file or return the error?  # noqa: TD003
+        files_parsed: dict[vn.Version, dict[str, str | dict[Any, Any]]] = {}
+        for f in files_known_format:
+            filename = f["filename"]
+            archive_ext = _parse_archive_extension(filename)
+            if isinstance(archive_ext, PyProjectProviderQueryError):
+                continue
+
+            v = _parse_version_from_filename(filename, name, archive_ext)
+
+            if v is None:
```

Comment:
Same question as above

Reply:
This would only happen if the sdist filename is not correctly formatted, which would violate the PyPI specification of sdist filenames. So I guess I can replace this with an assert.